### PR TITLE
Magnetic flux field equivalency

### PR DESCRIPTION
**Pull request by samjrholt**
### Description

Add functionality to Astropy to convert between two very common measures of magnetic field: magnetic flux density and magnetic field strength. The relation between these two quantities is straight forward in a vacuum:

```math
\mathbf{B} = \mu_0 \mathbf{H}
```

References: 
https://www.e-magnetica.pl/doku.php/confusion_between_b_and_h (Summary of equations)

Fixes #<Issue Number>

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

File: astropy/units/equivalencies.py
Comment (pllim) [inline astropy/units/equivalencies.py:486]: This change looks wrong.

**Comment by pllim** [comment]
Thanks! Definitely needs a change log but I'll let units maintainers decide if what's new also needed.

**Comment by pllim** [comment]
Can you also please edit the PR description to provide your motivation for this PR? Thanks!

File: astropy/units/equivalencies.py
Comment (samjrholt) [reply astropy/units/equivalencies.py]: Thank you for spotting this! It has now been reverted and corrected.

**Comment by mhvk** [comment]
Thanks for the PR. I guess it is the promised follow-up of #15962!

But like @pllim asked, it would still be useful to have a bit of a justification for adding the equivalency, especially as it is one that one can relatively easily do oneself, in a way that arguably is clearer for readers of the code (since they'll see an explicit division by `const.mu0`; though there are similar existing ones, like `temperature_energy`). Also, as I noted in #15962, it is not obvious this is something encountered frequently, and hence it is not obvious that it is worth the cost of making the list of equivalencies yet longer and therefore the likelihood that any equivalency gets found smaller...

Is the main benefit that one does not have to worry about working in SI or cgs in the conversion?

Note that if we pursue this, there definitely needs to be an entry in `docs/units/equivalencies.rst` explaining the use (this gets rendered in the documentation at https://docs.astropy.org/en/stable/units/equivalencies.html)

**Comment by samjrholt** [comment]
@mhvk you are correct this is the follow up to #15962! I have added a short justification in the PR description detailing the conversion with a reference. Let me know if you want me to extend it.

More than happy to have the conversation if it worth while to add this to the project. You are absolutely correct that this is a relatively straight forward conversion (simply a factor of $\mu_0$). From experience, there is often confusion between these two quantities even though they are strictly defined. A benefit that you mentioned is the SI/CGS conversion but also I think it should reduce user error/speed. If users need to do the simple conversion manually each time then they either need to memorize the equation or look it up. While this is a relatively trivial task for experienced users, we have had people in the past get this wrong due to the confusion that non-experts often have between B and H. I understand it is difficult to weigh it up against having too many but this convenience may be similar to the reason you have the other "simplistic" equivalencies such as `mass_energy` and `temperature_energy`?

Another option for a more complete version of the conversion is to have the full definition in matter

```math
\mathbf{B} = \mu_r \mu_0 \mathbf{H}
```
where `\mu_r` is the relative permeability. This can take a default value of 1 (the vacuum condition). However, while having more functionality it would likely be used less than the default version. 

I'll add to the `docs/units/equivalencies.rst` file while we decide if this is something you would like to support or not.

**Comment by samjrholt** [comment]
So there are a few converters online which I know colleagues often use to do this:

- https://maurermagnetic.com/en/demagnetizing/technology/convert-magnetic-units/
- https://www.kjmagnetics.com/unit.converter.asp
- https://www.bussi-demagnetizers.com/en/tecnologia-smagnetizzazione/italiano-convertitore-principali-unita-di-misura-del-magnetismo-e-del-magnetismo-residuo

File: astropy/units/equivalencies.py
Comment (nstarman) [reply astropy/units/equivalencies.py:879]: I agree. `mu_r != 1` hight be appreciated by the plasmapy folks and would better justify this Equivalence. With the `mu_r` set one could do conversions in a particular medium....

```python
# Calculate in a medium with mu_r = 0.9
with u.set_enabled_equivalencies(magnetic_flux_field(mu_r=0.9)):
    ...
```

File: docs/units/equivalencies.rst
Comment (samjrholt) [reply docs/units/equivalencies.rst:559]: I now have additional examples to which converts to CGS units.

File: astropy/units/equivalencies.py
Comment (samjrholt) [reply astropy/units/equivalencies.py]: I have added the parameter `mu_r` in the equivalency along with corresponding documentation and tests. Let me know if you would like anything changed :)

**Review by mhvk — APPROVED** [review]
Thanks, this looks all good! I'll wait a little with merging just in case of other comments.

**Review by nstarman — APPROVED** [review]
Thanks! Lgtm

File: astropy/units/equivalencies.py
Comment (pllim) [inline astropy/units/equivalencies.py:917]: I don't understand why codecov is complaining about this not covered.

File: docs/units/equivalencies.rst
Comment (pllim) [inline docs/units/equivalencies.rst:532]: Why is "Magnetic" repeated twice? Is this intentional?

File: astropy/units/equivalencies.py
Comment (pllim) [reply astropy/units/equivalencies.py:917]: `codecov/patch` says 100% so I would just ignore the complain here.

File: docs/units/equivalencies.rst
Comment (samjrholt) [reply docs/units/equivalencies.rst:532]:
```suggestion
Magnetic Flux Density and Magnetic Field Strength Equivalency
```
This was intentional but I understand it is not as readable! I have made a suggestion, alternatively could always drop the second magnetic as it is implied but I was erring on the side of caution.

**Comment by mhvk** [comment]
Thanks, I enabled auto-merged, so this should soon be in. Thanks, @samjrholt!

**Comment by neutrinoceros** [comment]
@mhvk Actually auto-merge might never complete: circleci is being awaited but was skipped because the last commit had `[skip CI]` in its message.

**Comment by pllim** [comment]
I squash-merged it. Thanks, all!

**Comment by samjrholt** [comment]
Thanks everyone!

**Pull request merged**
